### PR TITLE
[MODFIN-374]. Update budget validation when updating a budget

### PR DESCRIPTION
**Pull request by SerhiiNosko**
## Purpose
https://folio-org.atlassian.net/browse/MODFIN-374
Make validation the same as in mod-finance-storage BatchTransactionsChecks:

- https://github.com/folio-org/mod-finance-storage/blob/master/src/main/java/org/folio/service/transactions/batch/BatchTransactionChecks.java#L215

- https://github.com/folio-org/mod-finance-storage/blob/master/src/main/java/org/folio/service/transactions/batch/BatchTransactionChecks.java#L248

### CODE REVIEW

File: src/main/java/org/folio/services/budget/BudgetService.java [hunk 1]
```diff
@@ -146,26 +153,30 @@ private List<Error> checkRemainingEncumbrance(SharedBudget budget) {
...
-        .subtract(allocated.subtract(available.add(unavailable)))
-        .subtract(expenditures.subtract(credits).add(awaitingPayment))
-        .compareTo(BigDecimal.ZERO) < 0) {
+      BigDecimal allowableExpenditure = BigDecimal.valueOf(budget.getAllowableExpenditure()).movePointLeft(2);
+      BigDecimal totalFunding  = allocated.add(netTransfers);
+      // unavailable amount shouldn't be negative
+      BigDecimal unavailable = ensureNonNegative(awaitingPayment.add(expenditures).subtract(credits));
+
+      double remaining = totalFunding.multiply(allowableExpenditure).subtract(unavailable).doubleValue();
+      if (remaining < 0) {
```
Comment (imerabishvili) [inline src/main/java/org/folio/services/budget/BudgetService.java:171]: Maybe better to take without `doubleValue` and compareTo(ZERO)

File: src/main/java/org/folio/services/budget/BudgetService.java [hunk 2]
```diff
@@ -129,13 +129,20 @@ private List<Error> checkRemainingEncumbrance(SharedBudget budget) {
...
-    //[remaining amount we can encumber] = (allocated * allowableEncumbered) - (encumbered + awaitingPayment + expended)
+    // [remaining amount we can encumber] = (allocated + netTransfers) * allowableEncumbered - (encumbered + awaitingPayment + expended - credits)
+    // calculations should be consistent with BatchTransactionChecks in mod-finance-storage
     if (budget.getAllowableEncumbrance() != null) {
       log.info("checkRemainingEncumbrance:: Budget '{}' allowable encumbrance is not null", budget.getId());
-      BigDecimal newAllowableEncumbrance = BigDecimal.valueOf(budget.getAllowableEncumbrance()).movePointLeft(2);
-      if (totalFunding.multiply(newAllowableEncumbrance).compareTo(encumbered.add(awaitingPayment).add(expenditures).subtract(credits)) < 0) {
+      BigDecimal allowableEncumbrance = BigDecimal.valueOf(budget.getAllowableEncumbrance()).movePointLeft(2);
+      BigDecimal totalFunding = allocated.add(netTransfers);
+      // unavailable amount shouldn't be negative
```
Comment (imerabishvili) [inline src/main/java/org/folio/services/budget/BudgetService.java:141]: We can remove such comments when method names are self explanatory. E.g `zeroIfNegative`.

File: src/main/java/org/folio/services/budget/BudgetService.java [hunk 3]
```diff
@@ -146,26 +153,30 @@ private List<Error> checkRemainingEncumbrance(SharedBudget budget) {
...
+      if (remaining < 0) {
         log.error("checkRemainingExpenditure:: Allowable expenditure limit exceed for budget: {}", budget.getId());
         return Collections.singletonList(ALLOWABLE_EXPENDITURE_LIMIT_EXCEEDED.toError());
       }
     }
     return Collections.emptyList();
   }
+
+  private BigDecimal ensureNonNegative(BigDecimal amount) {
+    return amount.compareTo(BigDecimal.ZERO) < 0 ? BigDecimal.ZERO : amount;
```
Comment (imerabishvili) [inline src/main/java/org/folio/services/budget/BudgetService.java:180]: Same as `return amount.max(BigDecimal.ZERO);`

Comment (SerhiiNosko) [reply src/main/java/org/folio/services/budget/BudgetService.java]: thanks, applied suggestion

File: src/main/java/org/folio/services/budget/BudgetService.java (hunk 1 above)
Comment (SerhiiNosko) [reply src/main/java/org/folio/services/budget/BudgetService.java:171]: I thought about this, but decided to remain as in mod-finance-storage, to have exactly the same consistent code

https://github.com/folio-org/mod-finance-storage/blob/master/src/main/java/org/folio/service/transactions/batch/BatchTransactionChecks.java#L215

https://github.com/folio-org/mod-finance-storage/blob/master/src/main/java/org/folio/service/transactions/batch/BatchTransactionChecks.java#L248

File: src/main/java/org/folio/services/budget/BudgetService.java (hunk 2 above)
Comment (SerhiiNosko) [reply src/main/java/org/folio/services/budget/BudgetService.java:141]: agree, but remained with the same reason to keep the same code as in mod-finance-storage

**Review by azizbekxm — APPROVED** [review]

**Review by imerabishvili — APPROVED** [review]

**Pull request merged**